src/ostree/ot-admin-builtin-finalize-staged.c

```diff
@@ -51,10 +62,46 @@ ot_admin_builtin_finalize_staged (int argc, char **argv, OstreeCommandInvocation
   g_autoptr(GOptionContext) context = g_option_context_new ("");
   g_autoptr(OstreeSysroot) sysroot = NULL;
   if (!ostree_admin_option_context_parse (context, options, &argc, &argv,
-                                          OSTREE_ADMIN_BUILTIN_FLAG_SUPERUSER,
+                                          OSTREE_ADMIN_BUILTIN_FLAG_SUPERUSER | OSTREE_ADMIN_BUILTIN_FLAG_UNLOCKED,
                                           invocation, &sysroot, cancellable, error))
     return FALSE;
 
+  /* In case it's an automount, open /boot so that the automount doesn't expire
+   * until before this process exits. If it did expire and got unmounted, the
+   * service would be stopped and the deployment would be finalized earlier
+   * than expected.
+   */
+  int sysroot_fd = ostree_sysroot_get_fd (sysroot);
+  glnx_autofd int boot_fd = -1;
+  g_debug ("Opening /boot directory");
+  if (!glnx_opendirat (sysroot_fd, "boot", TRUE, &boot_fd, error))
+    return FALSE;
+
+  /* We want to wait until the system is shut down to actually finalize, so
+   * block on SIGTERM under the assumption that it will be received when
+   * systemd stops the unit.
+   */
+  gboolean running = TRUE;
+  g_unix_signal_add (SIGTERM, sigterm_cb, &running);
+  g_print ("Waiting for SIGTERM\n");
+  while (running)
+    g_main_context_iteration (NULL, TRUE);
+
+  /* Since the sysroot was initially started unlocked, setup a mount namespace
+   * to allow /boot and /sysroot to be writable, lock the sysroot, and reload
+   * it to ensure the state of the deployments is current.
+   *
+   * FIXME: This overlaps with the mount namespace handling in
+   * ostree_admin_option_context_parse. That should be factored out.
+   */
+  if (unshare (CLONE_NEWNS) < 0)
```

Comment:
Ohhh, right. I think it would be cleaner is to have systemd do this for us via e.g.
`MountFlags=slave`
in the unit.  See e.g.
https://github.com/coreos/rpm-ostree/blob/baa2eba2d77647204be2baff2b385e16220e0fac/src/daemon/rpm-ostreed.service.in#L11
(In rpm-ostree we always do everything via the daemon; unfortunately that's not true of ostree today, but in *this* case we know we're running in systemd and can use its features)

(Although I guess this direction conflicts with people who are trying to do ostree without systemd, but...I think the onus is going to be increasingly on them to at least parse a subset of unit files)

But, OK as is too - i.e. we can merge this and do followups.

Reply:
Just to make sure I'm following this right, because the unit has `ProtectHome` and `ReadOnlyPaths`, then systemd is already going to setup a mount namespace and there's no reason to create another one. Is that right? I sorta feel like you'd want to have code to check whether the process was in its own namespace or not.

Another thing I thought of while looking at this section. The unit talks about how only `ProtectHome` can be used, but then the sysroot goes and remounts `/boot` and `/sysroot` readwrite. Furthermore, it talks about how it needs to remove `/var/.updated`, but isn't that within the stateroot `/var`? Oh, I see that there's some handling for `/var` as a separate mount. Anways, this all looks like it could be handles with:

```
ProtectSystem=strict
ReadWritePaths=/var
```

with the mount handling. Maybe another day, though.

Reply:
Personally I've definitely called `ostree admin finalize-staged` manually when debugging stuff. I could fake out the `INVOCATION_ID` I guess, but maybe more elegant to do:
- if in systemd unit, don't unshare
- otherwise unshare or die trying

?